django/db/backends/postgresql/introspection.py

```diff
@@ -40,17 +40,20 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
 
     def get_field_type(self, data_type, description):
         field_type = super().get_field_type(data_type, description)
-        if description.is_autofield or (
-            # Required for pre-Django 4.1 serial columns.
-            description.default
-            and "nextval" in description.default
-        ):
+        if description.is_autofield:
             if field_type == "IntegerField":
                 return "AutoField"
             elif field_type == "BigIntegerField":
                 return "BigAutoField"
             elif field_type == "SmallIntegerField":
                 return "SmallAutoField"
+        elif description.default and "nextval" in description.default:
+            if field_type == "IntegerField":
+                return "SerialField"
+            elif field_type == "BigIntegerField":
+                return "BigSerialField"
+            elif field_type == "SmallIntegerField":
+                return "SmallSerialField"
```

Comment:
⚠️  There potential for breakage of existing projects here...

This check on default value was kept for backward compatibility as, prior to Django 4.1, the implementation of `AutoField` on PostgreSQL didn't use identity columns but serial columns with sequences. (See [here](https://docs.djangoproject.com/en/stable/releases/4.1/#models).)

This change will mean that all of those `AutoField`s that were created before Django 4.1, will now be treated as `SerialField`s and break a lot of stuff. We will need some way to distinguish between these cases. See [my comment](https://github.com/django/django/pull/15542#issuecomment-1403697290) which highlights that it wasn't necessary to migrate to identity columns, but that it was achievable with a manual migration. Many will have upgraded Django and not done this.

Reply:
I see, thanks for this @ngnpope . You're right, this is backwards incompatible.

I suppose the correct solution would be to force users to migrate first.

As a workaround, I could attach a comment to `SerialField`s storing the current Django version and check for this comment's presence on introspection. This is kind of a hack, as it would mean users can't attach a comment to `SerialField`s, but it would work as a temporary solution (unless someone has this very specific comment on their `AutoField` already, but I highly doubt that).

Reply:
> I suppose the correct solution would be to force users to migrate first.

I'm not sure that's viable. It would be a hard break for something that worked fine and there may be other ramifications if other non-Django code is doing things with those tables...

> ...but it would work as a temporary solution...

It's a case of break one feature to enable another. And if a DBA manually removed or changed a comment then the new feature would break. And at what point does this "temporary" solution get replaced with a permanent one?

I haven't had much time to dig into potential answers to this I'm afraid.

Reply:
Ok, in that case, we can leave it the way it is. The feature will work fine without this `get_field_type` function. This function is only used by the `inspectdb` command, which is why I was thinking it was okay to break backwards-compatibility in the first place.

Reply:
🙈 Ah, I was having an off day yesterday and missed that this was introspection-only.

Perhaps it is fine to make this change then as, while backward-incompatible, we don't make guarantees that introspection will give a perfect result.

In `django.core.management.commands.inspectdb.Command.get_field_types()` we have some `field_notes` that get output as comments in the generated code. It'd be nice to add a comment for `SerialField`, etc. that it may be intended that it is an `AutoField` or something if we were to go ahead with this change. I'm not sure how that'd work exactly as we probably don't want to reference `django.contrib.postgres` fields directly there though.

I would say that this caution still applies in general, however, e.g. we likely want to test your migration changes with a model that starts out as a pre-Django 4.1 `AutoField` for PostgreSQL to see how it behaves.

Reply:
I just realized `inspectdb` cannot output Postgres-specific fields. It will generate `models.SerialField` which is incorrect. Let me see if that's possible to fix...

Reply:
@ngnpope , I pushed the changes addressing this comment.